### ISSUE TITLE
fix how AbstractTrainer handles `validate` and `patience` for VOC dataset with different loss functions

### ISSUE THREAD

**Issue by NickleDave** [NickleDave/visual-search-nets#61]
right now the `AbstractTrainer.validate` method checks the number of dimensions of the target `y` and if it's greater than 1, it applies a sigmoid and then computes `f1`.

Otherwise it uses "accuracy" (true positive rate, recall).

but when training on PascalVOC with different loss functions, this could cause early stopping due to low f1 score based on sigmoid threshold, even when training with cross entropy to achieve accuracy on the largest object or on a random object

this in turn could cause these models to have artificially low accuracy. Unclear right now whether this is the case but it could be.

More importantly, the experiments are to test how diff't loss functions impact "behavior", so we want to make sure we do the best we can *under a given loss function*. So `patience` / early stopping should depend on the metric related to that loss function. I.e., if we use cross-entropy loss that computes accuracy for the largest object, then the `patience` should depend on accuracy for the largest object, not the f1 score for all objects.

At a high level, need to:
- log all metrics during validation / training for VOC
- have `patience` depend on metric used for loss function

In terms of implementation, need to:
- add a `loss_func` attribute (a string) to the `AbstractTrainer` that can be used for control flow statements in `validate` and `train` methods
- update `validate` method so that
  - [x] explicitly check whether the datset `isinstance` of `SearchStims` or `VOC`, and use that when deciding which accuracy to compute
  - [x] additionally for `VOC` it should compute "all" the accuracies, `BCE` --> f1, for any cross-entropy it should compute accuracy on random + largest objects (this will require a new `VOC` dataset object)
  - [x] return all those metrics in a dictionary that `train` can use
- update `train` method so that
  - [x] log all the metrics returned by `validate` to the `summary_writer`
  - [x] use the `dataset` and `loss_func` to determine which metric to use for `patienceå`
